msal_extensions/persistence.py

```diff
@@ -199,6 +216,14 @@ def load(self):
                 "DPAPI error likely caused by file content not previously encrypted. "
                 "App developer should migrate by calling save(plaintext) first.")
             raise
+        try:
+            return self._dp_agent.unprotect(data)
+        except OSError as exception:
+            raise PersistenceDecryptionError(
+                err_no=getattr(exception, "winerror", None),  # Exists in Python 3 on Windows
+                message="Decryption failed: {}. You may have to delete the file.".format(exception),  # pylint: disable=consider-using-f-string
```

Comment:
in .NET we delete it ourselves.

Reply:
Thanks for bringing up this intriguing topic.

The current "you may have to (manually) delete the file" advice was indeed influenced by the actual workaround that we gave to those few Azure CLI's end users who ran into this issue. So, when/if those end users took our advice, they implicitly consented and they knew what they were doing.

But it feels uncomfortable for our implementation to do this silently. The deletion is an irrevocable and potentially destructive operation.

Azure CLI team and us have not quite figured out why end users would run into this issue in the first place. Although the following speculation may or may not be the case for Azure CLI, but, if a downstream app would happen to use same filename for plaintext data and encrypted data, it will then be theoretically possible that a plaintext file predates the decryption attempt. In such case, a smooth recovery plan would be for the app developer to gracefully convert the plaintext data into encrypted data:

```python
encrypted_persistence = FilePersistenceWithDataProtection("my_data.bin")
try:
    valuable_data = encrypted_persistence.load()
except PersistenceDecryptionError:
    valuable_data = FilePersistence("my_data.bin").load()
    encrypted_persistence.save(valuable_data)
```

Or, if the app developer would still prefer, they can catch the `PersistenceDecryptionError` and do the file removal there. The decision is in their hands.

I incline to keep this PR's current behavior i.e. NOT automatically delete the file. We may, however, somehow reword the advice to "App developer or end user may choose to delete the file, or app developer could recover by blah blah blah". The "blah blah blah" part will be lengthy and the audience is the app developer. It is probably fine, as we anticipate the app developer to (eventually) catch this exception and translate it into a more user-friendly message to their end users.

Reply:
UPDATE: Now the exception message contains a link to this newly-created wiki page https://github.com/AzureAD/microsoft-authentication-extensions-for-python/wiki/PersistenceDecryptionError

Reply:
Azure CLI guarantees the encrypted file and unencrypted file do _not_ have the same name.

https://github.com/Azure/azure-cli/blob/fb805f4bb3167623107c10ec5f34581a838c8a1c/src/azure-cli-core/azure/cli/core/auth/persistence.py#L38

```py
file_extensions = {True: '.bin', False: '.json'}
...
    location += file_extensions[encrypt]
```